### ISSUE TITLE
[JENKINS-44699] Limit the number of lines that are shown for the log in the results page

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/blueocean-plugin#4183]
<p>Blue Ocean streams an exceptional amount of logs to the client. </p>

<p><b>In Scope</b></p>
<ul>
	<li>Limit all steps to show a maximum of 100 lines (last 100)</li>
</ul>

<p><b>Out of scope</b></p>
<ul>
	<li>Any functional changes to the amount of logs shown for non-pipeline</li>
</ul>

<p><b>Notes</b><br/>
There may be a frontend and backend component to this. </p>

<p>This means that there would only be 100 lines visible on screen at all times and the only way to see the whole log for the step is the "show more" button.</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay">jamesdumay</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44699" target="_blank">Limit the number of lines that are shown for the log in the results page</a></i></summary>
<i><ul>
<li><b>status</b>: Open
<li><b>priority</b>: Major
<li><b>component(s)</b>: blueocean-plugin
<li><b>resolution</b>: Unresolved
<li><b>votes</b>: 5
<li><b>watchers</b>: 11
<li><b>imported</b>: 2025-12-01
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Blue Ocean streams an exceptional amount of logs to the client. </p>

<p><b>In Scope</b></p>
<ul>
	<li>Limit all steps to show a maximum of 100 lines (last 100)</li>
</ul>

<p><b>Out of scope</b></p>
<ul>
	<li>Any functional changes to the amount of logs shown for non-pipeline</li>
</ul>

<p><b>Notes</b>
There may be a frontend and backend component to this. </p>

<p>This means that there would only be 100 lines visible on screen at all times and the only way to see the whole log for the step is the "show more" button.</p></pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44699?focusedId=309500&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-309500">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=vivek" class="user-hover" rel="vivek">vivek</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jmac11" class="user-hover" rel="jmac11">jmac11</a> it would be good if you two could collaborate on this. Some frontend work to change the behaviour of our log lines and some backend changes to the behaviour of the page of logs we get back to server.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=vivek" class="user-hover" rel="vivek">vivek</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jmac11" class="user-hover" rel="jmac11">jmac11</a> it would be good if you two could collaborate on this. Some frontend work to change the behaviour of our log lines and some backend changes to the behaviour of the page of logs we get back to server.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/vivek.png" title="vivek's avatar" /> vivek:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44699?focusedId=310526&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-310526">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Serving last n number of lines could be expensive, we can end up reading all file. Instead Logging API for step or for run serves last 150KB log data by default. It's controllable using query param thresholdInKB, its default value is 150. Basically it results in to computing offset,  `offset=logFileLength-thresholdInKB` in to the log file.</p>

<p>For example, this is default rest API called from UI, it fetches last 150KB of log, in terms of line this particular log is about 1900 lines.</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">https:<span class="code-comment">//ci.blueocean.io/blue/<span class="code-keyword">rest</span>/organizations/jenkins/pipelines/blueocean/branches/bug%252FJENKINS-45754-github-enterprise-url-validation-2/runs/2/nodes/35/steps/38/log/</span>
</pre>
</div></div>

<p>with threasholdInKB=1, we get about 10 lines:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">https:<span class="code-comment">//ci.blueocean.io/blue/<span class="code-keyword">rest</span>/organizations/jenkins/pipelines/blueocean/branches/bug%252FJENKINS-45754-github-enterprise-url-validation-2/runs/2/nodes/35/steps/38/log/?thresholdInKB=1</span>
</pre>
</div></div>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">ESS [04:52 min]
[INFO] Events API <span class="code-keyword">for</span> Blue Ocean .......................... SUCCESS [ 41.221 s]
[INFO] Dashboard <span class="code-keyword">for</span> Blue Ocean ........................... SUCCESS [03:37 min]
[INFO] Personalization <span class="code-keyword">for</span> Blue Ocean ..................... SUCCESS [02:27 min]
[INFO] Config API <span class="code-keyword">for</span> Blue Ocean .......................... SUCCESS [01:03 min]
[INFO] GitHub Pipeline <span class="code-keyword">for</span> Blue Ocean ..................... SUCCESS [03:47 min]
[INFO] Git Pipeline <span class="code-keyword">for</span> Blue Ocean ........................ SUCCESS [01:08 min]
[INFO] Bitbucket Pipeline <span class="code-keyword">for</span> Blue Ocean .................. SUCCESS [03:20 min]
[INFO] Blue Ocean ......................................... SUCCESS [ 50.354 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 30:21 min
[INFO] Finished at: 2017-08-15T20:21:04+00:00
[INFO] Final Memory: 174M/1599M
[INFO] ------------------------------------------------------------------------
</pre>
</div></div>

</pre>
</details>

</ul>
</details>

<p>Serving last n number of lines could be expensive, we can end up reading all file. Instead Logging API for step or for run serves last 150KB log data by default. It's controllable using query param thresholdInKB, its default value is 150. Basically it results in to computing offset,  `offset=logFileLength-thresholdInKB` in to the log file.</p>

<p>For example, this is default rest API called from UI, it fetches last 150KB of log, in terms of line this particular log is about 1900 lines.</p>


<pre>
https:<span class="code-comment">//ci.blueocean.io/blue/<span class="code-keyword">rest</span>/organizations/jenkins/pipelines/blueocean/branches/bug%252FJENKINS-45754-github-enterprise-url-validation-2/runs/2/nodes/35/steps/38/log/</span>
</pre>

<p>with threasholdInKB=1, we get about 10 lines:</p>

<pre>
https:<span class="code-comment">//ci.blueocean.io/blue/<span class="code-keyword">rest</span>/organizations/jenkins/pipelines/blueocean/branches/bug%252FJENKINS-45754-github-enterprise-url-validation-2/runs/2/nodes/35/steps/38/log/?thresholdInKB=1</span>
</pre>


<pre>
ESS [04:52 min]
[INFO] Events API <span class="code-keyword">for</span> Blue Ocean .......................... SUCCESS [ 41.221 s]
[INFO] Dashboard <span class="code-keyword">for</span> Blue Ocean ........................... SUCCESS [03:37 min]
[INFO] Personalization <span class="code-keyword">for</span> Blue Ocean ..................... SUCCESS [02:27 min]
[INFO] Config API <span class="code-keyword">for</span> Blue Ocean .......................... SUCCESS [01:03 min]
[INFO] GitHub Pipeline <span class="code-keyword">for</span> Blue Ocean ..................... SUCCESS [03:47 min]
[INFO] Git Pipeline <span class="code-keyword">for</span> Blue Ocean ........................ SUCCESS [01:08 min]
[INFO] Bitbucket Pipeline <span class="code-keyword">for</span> Blue Ocean .................. SUCCESS [03:20 min]
[INFO] Blue Ocean ......................................... SUCCESS [ 50.354 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 30:21 min
[INFO] Finished at: 2017-08-15T20:21:04+00:00
[INFO] Final Memory: 174M/1599M
[INFO] ------------------------------------------------------------------------
</pre>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>sophistifunk:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44699?focusedId=310792&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-310792">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<ol>
	<li>What's our main concern here? Reducing the number of lines we see, or network traffic?</li>
	<li>Do we want to discard lines at the front-end if we end up fetching more than the threshold because they're short?</li>
	<li>Is there a particular benefit to having a different line count for live vs completed steps to justify the complexity? If so, what is it, and do we want to discard 50 lines when the step you're watching is complete?</li>
	<li>Do the line numbers in the gutter have any value to the user when we don't have the entire log for that step? Unless we've fetched the entire thing, they're going to be wrong, and will change when the user gets more data, either by requesting more or watching a live log.</li>
</ol>

</pre>
</details>

</ul>
</details>

<ol>
	<li>What's our main concern here? Reducing the number of lines we see, or network traffic?</li>
	<li>Do we want to discard lines at the front-end if we end up fetching more than the threshold because they're short?</li>
	<li>Is there a particular benefit to having a different line count for live vs completed steps to justify the complexity? If so, what is it, and do we want to discard 50 lines when the step you're watching is complete?</li>
	<li>Do the line numbers in the gutter have any value to the user when we don't have the entire log for that step? Unless we've fetched the entire thing, they're going to be wrong, and will change when the user gets more data, either by requesting more or watching a live log.</li>
</ol>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44699?focusedId=310793&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-310793">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Good questions</p>
<ol>
	<li>The driver is that are displaying too much useless information on screen. The last 100-150 lines contains the context.</li>
	<li>Yes, that would prevent things from jumping around too I suspect.</li>
	<li>Lets kill the line numbers. They just don't make any sense.</li>
	<li>Ditto. Eventually we should have a timestamp but I suspect that is going to be hard (I suspect core changes, <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=vivek" class="user-hover" rel="vivek">vivek</a>?)</li>
</ol>


<p>EDIT: lets simplify here. Lets stick with 100 lines for both running and finalized log steps.</p>
</pre>
</details>

</ul>
</details>

<p>Good questions</p>
<ol>
	<li>The driver is that are displaying too much useless information on screen. The last 100-150 lines contains the context.</li>
	<li>Yes, that would prevent things from jumping around too I suspect.</li>
	<li>Lets kill the line numbers. They just don't make any sense.</li>
	<li>Ditto. Eventually we should have a timestamp but I suspect that is going to be hard (I suspect core changes, <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=vivek" class="user-hover" rel="vivek">vivek</a>?)</li>
</ol>


<p>EDIT: lets simplify here. Lets stick with 100 lines for both running and finalized log steps.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>sophistifunk:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44699?focusedId=311393&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-311393">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Do we want the users to be able to request "more" log, or just "all" the log if they need more info? My vote is on "all" for implementation simplicity and fewer choices the user has to make, but I thought I should clarify that.</p>
</pre>
</details>

</ul>
</details>

<p>Do we want the users to be able to request "more" log, or just "all" the log if they need more info? My vote is on "all" for implementation simplicity and fewer choices the user has to make, but I thought I should clarify that.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44699?focusedId=311396&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-311396">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=sophistifunk" class="user-hover" rel="sophistifunk">sophistifunk</a> I believe the "all" behaviour is what we have today. I agree, lets keep it that way unless we get a strong signal to change it.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=sophistifunk" class="user-hover" rel="sophistifunk">sophistifunk</a> I believe the "all" behaviour is what we have today. I agree, lets keep it that way unless we get a strong signal to change it.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44699?focusedId=316678&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-316678">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=sophistifunk" class="user-hover" rel="sophistifunk">sophistifunk</a> any moar progress on this or should we take this out of the queue? </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=sophistifunk" class="user-hover" rel="sophistifunk">sophistifunk</a> any moar progress on this or should we take this out of the queue? </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>sophistifunk:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44699?focusedId=316680&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-316680">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>There is some progress, it's bit-rotting in a local branch (slowly though, it's mostly new files) while waiting on some fixes <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kzantow" class="user-hover" rel="kzantow">kzantow</a> is sitting on <b><span class="error">&#91;cough&#93;</span></b></p>
</pre>
</details>

</ul>
</details>

<p>There is some progress, it's bit-rotting in a local branch (slowly though, it's mostly new files) while waiting on some fixes <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kzantow" class="user-hover" rel="kzantow">kzantow</a> is sitting on <b><span class="error">&#91;cough&#93;</span></b></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44699?focusedId=316681&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-316681">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>np - just put this one back behind <a href="https://issues.jenkins-ci.org/browse/JENKINS-45182" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-45182</a></p>
</pre>
</details>

</ul>
</details>

<p>np - just put this one back behind <a href="https://issues.jenkins-ci.org/browse/JENKINS-45182" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-45182</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>sophistifunk:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44699?focusedId=317809&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-317809">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Back onto this now</p>
</pre>
</details>

</ul>
</details>

<p>Back onto this now</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>gluszczykk:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44699?focusedId=408278&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-408278">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=sophistifunk" class="user-hover" rel="sophistifunk">sophistifunk</a> any update on this <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/>?</p>
</pre>
</details>

</ul>
</details>

<p>Hi <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=sophistifunk" class="user-hover" rel="sophistifunk">sophistifunk</a> any update on this <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/>?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>sophistifunk:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44699?focusedId=408303&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-408303">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Sorry I'm not currently on the Jenkins team, your best bet will be the mailing lists.</p>
</pre>
</details>

</ul>
</details>

<p>Sorry I'm not currently on the Jenkins team, your best bet will be the mailing lists.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-44699?focusedId=408306&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-408306">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=gluszczykk" class="user-hover" rel="gluszczykk">gluszczykk</a> it is unlikely this will be implemented unless a Jenkins community member does the implementation.  See the <a href="https://groups.google.com/g/jenkinsci-dev/c/zO9VFNxhF-o/m/EpjdA8-jCgAJ" class="external-link" target="_blank" rel="nofollow noopener">Jeremy Hartley's message</a> for the latest status on new feature development in Blue Ocean.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=gluszczykk" class="user-hover" rel="gluszczykk">gluszczykk</a> it is unlikely this will be implemented unless a Jenkins community member does the implementation.  See the <a href="https://groups.google.com/g/jenkinsci-dev/c/zO9VFNxhF-o/m/EpjdA8-jCgAJ" class="external-link" target="_blank" rel="nofollow noopener">Jeremy Hartley's message</a> for the latest status on new feature development in Blue Ocean.</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `is_duplicated_by` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-35864%22">JENKINS-35864</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `is_duplicated_by` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-44327%22">JENKINS-44327</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `is_duplicated_by` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-46442%22">JENKINS-46442</a>]</i>
